### ISSUE TITLE
PureCN error

### ISSUE THREAD

**Issue by Shenglai** [lima1/PureCN#107]
Hi @lima1 

I'm having an issue running with the latest stable version of PureCN (PureCN_1.14.3).

It looks like to an old issue https://github.com/lima1/PureCN/issues/20#issuecomment-368330032

My error is:
```
INFO [2019-10-09 19:19:30] Fitting variants for local optimum 16/16...
INFO [2019-10-09 19:19:34] Fitting variants for purity 0.39, tumor ploidy 2.87 and contamination 0.01.
INFO [2019-10-09 19:19:50] Poor goodness-of-fit (0.720). Skipping post-optimization.
INFO [2019-10-09 19:19:50] Optimized purity: 0.39
INFO [2019-10-09 19:19:51] Done.
INFO [2019-10-09 19:19:51] ------------------------------------------------------------
Warning message:
In .bcfHeaderAsSimpleList(header) :
  duplicate keys in header will be forced to unique rownames
null device
          1
null device
          1
null device
          1
[job purecn] Job error:
```
Attached is the log from `PureCN.R`, but it seems it completes the major part, and error is not shown in the log. 
[3b036cbf-b0e7-4fd3-8b22-734ec845e263.log](https://github.com/lima1/PureCN/files/3709242/3b036cbf-b0e7-4fd3-8b22-734ec845e263.log)

Please check the following json to see my inputs and comments.
```
{
    "thread_num": 4,
    "intervalweightfile": {
        "path": "interval_weights.sureselectxthumanallexonv5.tcga-meso.hg38.txt",
        "class": "File",
        "__comments": "This is generated from PureCN/extdata/NormalDB.R with NormalDB."
    },
    "normaldb": {
        "path": "normalDB.sureselectxthumanallexonv5.tcga-meso.hg38.rds",
        "class": "File",
        "__comments": "This is generated from PureCN/extdata/NormalDB.R with IntervalWeight."
    },
    "genome": "hg38",
    "fasta": {
        "path": "GRCh38.d1.vd1.fa",
        "class": "File",
        "__comments": "GDC reference."
    },
    "bigwig": {
        "path": "GRCh38.d1.vd1.100.bigWig",
        "class": "File",
        "__comments": "Mappability file."
    },
    "exclude": {
        "path": "GRCh38.d1.vd1.centromeres.telomeres.purecn.bed",
        "class": "File",
        "__comments": "GRCh38 centromerers and telomeres regions bed file. For IntervalFile.R"
    },
    "capture_kit": {
        "path": "S04380110_Covered.bed.gz",
        "class": "File",
        "__comments": "Baits bed for SureSelect Human All Exon v5."
    },
    "sampleid": "3b036cbf-b0e7-4fd3-8b22-734ec845e263",
    "bam": {
        "path": "3b036cbf-b0e7-4fd3-8b22-734ec845e263_gdc_realn.bam",
        "class": "File"
    },
    "vcf": {
        "path": "bb5b9508-7a40-4851-bec8-5d168ca4a07e.wxs.tumor_only.gatk4_mutect2.raw_somatic_mutation.vcf.gz",
        "class": "File",
        "__comments": "GATK4.1.2 Mutect2 tumor only calling. Followed by Broad's best practices. https://github.com/broadinstitute/gatk/tree/master/scripts/mutect2_wdl"
    }
}
```
CMD for PureCN.R:
```
Rscript \
/usr/local/lib/R/site-library/PureCN/extdata/PureCN.R \
--cores \
4 \
--force \
--genome \
hg38 \
--intervals \
GRCh38.d1.vd1.S04380110_Covered.hg38.txt \
--normaldb \
normalDB.sureselectxthumanallexonv5.tcga-meso.hg38.rds \
--out \
. \
--outvcf \
--postoptimize \
--sampleid \
3b036cbf-b0e7-4fd3-8b22-734ec845e263 \
--tumor \
3b036cbf-b0e7-4fd3-8b22-734ec845e263_gdc_realn_coverage_loess.txt \
--vcf \
bb5b9508-7a40-4851-bec8-5d168ca4a07e.wxs.tumor_only.gatk4_mutect2.raw_somatic_mutation.vcf.gz
```

The exclude bed was added after I check'ed the old issue #20 , and saw you commenting about the concern on centromere regions. I would get exactly the same error if I ran the purecn w/o this bed file. (It makes sense since the outputs' md5sums from `IntervalFile.R`, w/ or w/o exclude bed, are identical.)

I also tried the debug method in this comment, https://github.com/lima1/PureCN/issues/20#issuecomment-368347755. However, I didn't see the line `"centromeres <- .getCentromeres(res)"` after `PureCN:::.getArmLocations(x)`, so I was just able to press enter until it errors out. The error is: 
```
debug: chromCoords <- suppressWarnings(t(vapply(split(start(x), as.character(seqnames(x))), 
    function(y) c(min(y), max(y)), c(min = double(1), max = double(1)))))
Browse[2]> 
Error in (function (classes, fdef, mtable)  : 
  unable to find an inherited method for function ‘seqnames’ for signature ‘"list"’
```

I also check'ed this issue #79 . So I also tried running the latest PureCN (PureCN_1.15.24). However, I'm still getting the same error. (It seems to me that the GATK4 PRs are merged to the master branch, but I could be wrong.)

More information that I could think of: 
1. The NormalDB was created by 39 bam files. I'm not sure if it is a big enough pool of samples. 
2. I can not run `calculateMappingBiasVcf` now for new GATK4 PON, since the PON VCF now automatically removes all the sample columns, and `calculateMappingBiasVcf` will complain the pon only contains a single sample. I'm not sure if adding a mapping bias rds file could resolve the issue or not.

In summary, I will have the following error, on stable/dev PureCN version (1.14.3, and 1.15.24), w/ or w/o a exclude bed file.

```
Warning message:
In .bcfHeaderAsSimpleList(header) :
  duplicate keys in header will be forced to unique rownames
null device
          1
null device
          1
null device
          1
```
I also created a `_bestonly.rds` followed by https://github.com/lima1/PureCN/issues/20#issuecomment-368336340 for debug purpose. (It doesn't fail at `callLOH()`, just FYI.) 
[3b036cbf-b0e7-4fd3-8b22-734ec845e263_bestonly.rds.zip](https://github.com/lima1/PureCN/files/3709403/3b036cbf-b0e7-4fd3-8b22-734ec845e263_bestonly.rds.zip)

**Comment by lima1** [comment]
Hi @Shenglai 

Looks good to me. Not sure what the error is or could be. What output files do you get?

You can also run R --args --rds 3b036cbf-b0e7-4fd3-8b22-734ec845e263_bestonly.rds.zip.
and then source("~/git/PureCN/inst/extdata/PureCN.R", echo=TRUE)

This should show you where it fails.

And yes, I don't think there is a good way yet of generating the PoN with GATK4. But I would highly recommend it, the FAQ section of the main vignette  shows how to do that with GATK3.

**Comment by lima1** [comment]
Ah, sorry, just saw the error you found. I’ll have a look on Friday.

**Comment by Shenglai** [comment]
```
You can also run R --args --rds 3b036cbf-b0e7-4fd3-8b22-734ec845e263_bestonly.rds.zip.
and then source("~/git/PureCN/inst/extdata/PureCN.R", echo=TRUE)

This should show you where it fails.
```

So I followed this, and it shows me where all the `null device 1` are coming from but not the error. 
```
> ### Run PureCN ----------------------------------------------------------------
>
> if (file.exists(file.rds) && !opt$force) {
+     flog.info("%s  ..." ... [TRUNCATED]
INFO [2019-10-10 04:00:30] 3b036cbf-b0e7-4fd3-8b22-734ec845e263.rds already exists. Skipping... (--force will overwrite)
INFO [2019-10-10 04:00:30] Reading 3b036cbf-b0e7-4fd3-8b22-734ec845e263.rds...
> ### Create output files -------------------------------------------------------
>
> curationFile <- createCurationFile(file.rds)
> if (debug) {
+     curationFile$log.ratio.offset <- mean(ret$results[[1]]$log.ratio.offset)
+     curationFile$log.ratio.sdev <- ret$input$log.ratio .... [TRUNCATED]
> file.pdf <- paste0(out, ".pdf")
> pdf(file.pdf, width = 10, height = 11)
> plotAbs(ret, type = "all")
> dev.off()
null device
          1
> file.pdf <- paste0(out, "_local_optima.pdf")
> pdf(file.pdf, width = 5, height = 5)
> plotAbs(ret, 1, type = "overview")
> dev.off()
null device
          1
> file.seg <- paste0(out, "_dnacopy.seg")
> seg <- ret$results[[1]]$seg
> seg <- seg[, c(1:6, match("C", colnames(seg)))]
> write.table(seg, file = file.seg, sep = "\t", quote = FALSE,
+     row.names = FALSE)
> if (is(ret$results[[1]]$gene.calls, "data.frame")) {
+     file.genes <- paste0(out, "_genes.csv")
+     allAlterations <- callAlterations(ret, all. .... [TRUNCATED]
> if (!is.null(ret$input$vcf)) {
+     if (opt$outvcf) {
+         file.vcf <- paste0(out, ".vcf")
+         vcfanno <- predictSomatic(ret, return.vcf .... [TRUNCATED]
null device
          1
```
But I guess 
```
debug: chromCoords <- suppressWarnings(t(vapply(split(start(x), as.character(seqnames(x))), 
    function(y) c(min(y), max(y)), c(min = double(1), max = double(1)))))
Browse[2]> 
Error in (function (classes, fdef, mtable)  : 
  unable to find an inherited method for function ‘seqnames’ for signature ‘"list"’
``` 
This is the real error we are looking for? Anyway, thanks for the reply.

**Comment by lima1** [comment]
I think the error you found is unrelated - in fact there is no error: the correct call of this function would be now: 

`PureCN:::.getArmLocations(x$input$vcf, centromeres = x$input$centromeres, chr.hash = PureCN:::.getChrHash(seqlevels(x$input$vcf)))
`
And this works fine.

I added some code to hide the null device output messages. But I think all the output files are generated. So not sure what's causing the error.

**Comment by Shenglai** [comment]
Thanks, I will try the latest one and try it on Monday.

**Comment by Shenglai** [comment]
Ah, I think you are right. It's not a PureCN error. `echo $?` returns 0 after running PureCN. I fixed it on my end now. Thanks for the help, and sorry for the confusion.

**Comment by Shenglai** [comment]
FYI, the error is from the part of CWL-docker.

**Comment by lima1** [comment]
Ok. Current GitHub will become the new stable in 2 weeks and has a few cleanups to API and output files. You probably want to start from there if you setup a pipeline now (instead of 1.14.3).  Not planning any more code changes besides bugfixes and documentation until release.

Also have a look at the PSCBS segmentation (search for it here: https://bioconductor.org/packages/devel/bioc/vignettes/PureCN/inst/doc/Quick.html). It will become the new default in the Spring 2020 release.

**Comment by Shenglai** [comment]
Thanks, good to know.

**Comment by Shenglai** [comment]
I also found that the output (just check'ed output VCFs) from PureCN is not deterministic. I had some slight differences between different runs with identical parameters. Is it expected?

**Comment by lima1** [comment]
Yes, it should be very minor, but yes, the optimization is stochastic (Simulated Annealing). You can add a seed in PureCN.R with --seed, that should fix this.
